### PR TITLE
Further refactoring around monitoring cabal files

### PR DESCRIPTION
**Pull request by dcoutts**
This builds on the partial fix from #3325. That patch makes sure that the content of the .cabal files is monitored.

In fact if the original code were working properly the conent would already be being monitored, and not just the content but also the search for the file in the case of a file glob (ie detecting changes in the set of files that match the glob).

So these patches fix the original problem plus some refactoring to eliminate the class of bug at work here. We do not revert the basic idea from #3325 however, we follow the idea that the code that searches for the files should monitor existence of the set of files that match the glob, while the code that reads the files should monitor the file content. So we keep the bit of #3325 that adds monitoring of the file content, and remove file content monitoring from the glob, keeping just monitoring file existence.

This ought to merge into 1.24 without too much problem. It's probably wise to do so otherwise when people use globs like: `packages: ./*/*.cabal` (which indeed is the default/implicit project config) then we'll fail to spot new files matching the glob (e.g. as a result of cabal unpack), since the fix in #3325 only covers the content of files we did find, not the glob.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
LGTM.

File: cabal-install/Distribution/Client/Glob.hs [hunk 1]
```diff
@@ -55,8 +55,7 @@ data GlobPiece = WildCard
 
 data FilePathRoot
    = FilePathRelative
-   | FilePathUnixRoot
-   | FilePathWinDrive Char
+   | FilePathRoot String  -- e.g. "/", "c:\" or result of FilePath.takeDrive
```
Comment (ezyang) [inline cabal-install/Distribution/Client/Glob.hs]: This is very minor, but why not a `FilePath` here?

File: cabal-install/Distribution/Client/ProjectConfig.hs [hunk 2]
```diff
@@ -656,9 +656,14 @@ findProjectPackages projectRootDir ProjectConfig{..} = do
                       && takeExtension (dropExtension f) == ".tar"
 
 
-globStarDotCabal :: FilePathGlob
-globStarDotCabal =
-    FilePathGlob FilePathRelative (GlobFile [WildCard, Literal ".cabal"])
+globStarDotCabal :: FilePath -> FilePathGlob
```
Comment (ezyang) [inline cabal-install/Distribution/Client/ProjectConfig.hs]: Sorry... what does this function do again?

**Comment by ezyang** [comment]
OK LGTM.

File: cabal-install/Distribution/Client/Glob.hs (hunk 1 above)
Comment (dcoutts) [inline cabal-install/Distribution/Client/Glob.hs]: Oh yes, should be.

File: cabal-install/Distribution/Client/ProjectConfig.hs (hunk 2 above)
Comment (dcoutts) [inline cabal-install/Distribution/Client/ProjectConfig.hs]: Will add a comment. It's $dir/*.cabal

**Comment by dcoutts** [comment]
Ok, will squash the fixes in and merge.

@23Skidoo would you mind considering this for 1.24 too. As explained at the top, I think the impact is that new .cabal files dropped in to match a glob will not be picked up, affecting the use case of `cabal get somedep`.

**Comment by 23Skidoo** [comment]
Sure, this can go into 1.24.

**Pull request merged**